pkg/controller/dns/external_dns_crd.go

```diff
@@ -0,0 +1,118 @@
+	"github.com/Azure/aks-app-routing-operator/pkg/util"
+	"github.com/go-logr/logr"
+	"github.com/hashicorp/go-multierror"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var ExternalDNSCRDControllerName = controllername.New("externaldns", "crd")
+
+type ExternalDNSCRDController struct {
+	config *config.Config
+	client client.Client
+	events record.EventRecorder
+}
+
+func NewExternalDNSCRDController(manager ctrl.Manager, config config.Config) error {
+	return ExternalDNSCRDControllerName.AddToController(ctrl.NewControllerManagedBy(manager).
+		For(&v1alpha1.ExternalDNS{}).
+		Owns(&appsv1.Deployment{}).
+		Owns(&corev1.ConfigMap{}), manager.GetLogger()).
+		Complete(&ExternalDNSCRDController{
+			config: &config,
+			client: manager.GetClient(),
+			events: manager.GetEventRecorderFor("aks-app-routing-operator"),
+		})
+}
+
+func (e *ExternalDNSCRDController) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	defer func() {
+		metrics.HandleControllerReconcileMetrics(ExternalDNSCRDControllerName, res, err)
+	}()
+
+	// set up logger
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("creating logger: %w", err)
+	}
+	logger = ExternalDNSCRDControllerName.AddToLogger(logger).WithValues("namespace", req.Namespace, "name", req.Name)
+
+	obj := &v1alpha1.ExternalDNS{}
+	if err = e.client.Get(ctx, req.NamespacedName, obj); err != nil {
+		logger.Error(client.IgnoreNotFound(err), "failed to get externaldns object, will ignore not found error")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// verify serviceaccount
+	if _, err = keyvault.GetServiceAccountAndVerifyWorkloadIdentity(ctx, e.client, obj.GetInputServiceAccount(), obj.GetNamespace()); err != nil {
+
+		var userErr keyvault.UserError
+		if errors.As(err, &userErr) {
+			logger.Info("failed to verify service account due to user error, sending warning event: " + userErr.UserError())
+			e.events.Eventf(obj, corev1.EventTypeWarning, "FailedUpdateOrCreateExternalDNSResources", "failed serviceaccount verification: %s", userErr.UserError())
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to verify service account")
+		return ctrl.Result{}, err
+	}
+
+	inputDNSConf := buildInputDNSConfig(obj)
+	manifestsConf, err := manifests.NewExternalDNSConfig(e.config, inputDNSConf)
+	if err != nil {
+		logger.Error(err, "failed to generate ExternalDNS resources from ExternalDNS CR")
+		e.events.Eventf(obj, corev1.EventTypeWarning, "FailedUpdateOrCreateExternalDNSResources", "failed to generate external DNS resources: %s", err)
+		return ctrl.Result{}, err
+	}
+
+	resources := manifestsConf.Resources()
+
+	multiError := &multierror.Error{ErrorFormat: func(errs []error) string {
```

Comment:
what's the reason for customizing the errorformat here? looks like the default one to me